### ISSUE TITLE
Tracking issuse: declarative macros 2.0

### ISSUE THREAD

**Issue by nrc** [rust-lang/rust#39412]
Tracking issue for declarative macros 2.0 (aka `macro_rules` aka macros-by-example).

RFC: TODO

RFC PR: https://github.com/rust-lang/rfcs/pull/1584

cc @rust-lang/compiler

**Comment by CryZe** [comment]
@nrc Typo in Issue Name: "issuse"

**Comment by jseyfried** [comment]
## Tasks
 - [ ] Complete the draft hygiene [prototype](https://github.com/jseyfried/rust/tree/hygiene) and [RFC](https://github.com/jseyfried/rfcs/blob/hygiene/text/0000-hygiene.md).
   - [x] Hygiene for items, explicit imports, lexical scopes, and module scopes.
   - [x] Hygiene for globs and trait methods (w.r.t. extension trait candidates in scope for a method call).
   - [x] Hygiene for type directed name resolutions (i.e. methods, associated types, and fields).
   - [ ] Support `macro` in blocks as well as modules.
   - [ ] Implement "hygiene bending" for when users want a name from macro def to "escape".
   - [ ] Implement inter-crate hygiene, except for nested `macro`s.
   - [ ] Implement inter-crate hygiene for nested `macro`s (pending macro def encoding).
   - [ ] Make `private_in_public` hygienic.
   - [ ] Make `unsafe` and lints hygienic.
 - [ ] Add variant `ast::ItemKind::MacroDef` for `macro_rules!` items and `macro` items.
 - [ ] Encode `macro`s in the crate metadata using `TokenStream`, not `String`.
 - [ ] Fix #30476 (`private_in_public` details).
 - [ ] Fix #30506 and #39450 (`Span` issues).
 - [ ] Land `macro` behind a feature gate.
 - [ ] Future-proof matchers (e.g. `$e:expr`) by employing a simpler, more general grammar.
 - [ ] Allow fragments (e.g. `$e` where `$e:expr`) to be parsed in more contexts (c.f. #26361).
 - [ ] Decide whether we want macro invocations in identifier positions and/or eager expansion.

**Comment by nrc** [comment]
We need to RFC a whole bunch of stuff here. In particular, I would like to propose some new syntax for declaring macros and we should RFC the changes to matchers.

**Comment by tikue** [comment]
Can the hygiene RFC mention pattern hygiene? This in particular scares me:

```rust
// Unsuspecting user's code
#[allow(non_camel_case_types)]
struct i(i64);

macro_rules! ignorant_macro {
    () => {
        let i = 0;
        println!("{}", i);
    };
}

fn main() {
    // oh no!
    ignorant_macro!();
}
```

**Comment by jseyfried** [comment]
@tikue I'm not sure patterns need special treatment with respect to hygiene.

For example, on the hygiene prototype,
```rust
#[allow(non_camel_case_types)]
struct i(i64);

macro ignorant_macro() {
    let i = 0; // ERROR: let bindings cannot shadow tuple structs
    println!("{}", i); 
}

fn main() {
    ignorant_macro!(); // NOTE: in this macro invocation
}
```

This makes sense to me since `let i = 0;` is shadowing a tuple struct.
In particular, if `let i = 0;` were removed then the following use of `i` would resolve to the tuple struct, [no matter where `ignorant_macro` is used](https://github.com/jseyfried/rfcs/blob/hygiene/text/0000-hygiene.md#names-in-macro-definitions-behave-reliably). 

Note the symmetry to [this example](https://play.rust-lang.org/?gist=056bf1634288336edd6aaa9f06595894&version=nightly&backtrace=0):
```rust
#[allow(non_camel_case_types)]
struct i(i64);

fn ignorant_fn() {
    let i = 0; // ERROR: let bindings cannot shadow tuple structs
    println!("{}", i); 
}
```
If the tuple struct `i` isn't in scope at `macro ignorant_macro() { ... }`, then the `let i = 0;` does not shadow it and there is no error. For example, the following compiles on the hygiene prototype:
```rust
mod foo {
    pub macro ignorant_macro() {
        let i = 0;
        println!("{}", i); // Without `let i = 0;`, there would be no `i` in scope here.
    }
}

// Unsuspecting user's code
#[allow(non_camel_case_types)]
struct i(i64);

fn main() {
    foo::ignorant_macro!();
}
```

I added these examples to the hygiene prototype's [test suite](https://github.com/rust-lang/rust/compare/master...jseyfried:hygiene#diff-3de01eaa403a16d51e1b6da505d15044).

**Comment by alexreg** [comment]
Has there been any progress on Macros 2.0 lately?

**Comment by alexreg** [comment]
For one, I was wondering whether repeated arguments are supported yet, as they are for `macro_rules`. If so, what is the syntax?

**Comment by mark-i-m** [comment]
Note for those who haven't seen yet: macros 2.0 is apparently slated to be _stable_ later this year, according to the proposed roadmap (rust-lang/rfcs#2314)...

On the one hand, that's pretty exciting :tada:!

On the other hand, I was really surprised that the feature is so close to being done and so little is known about  it by the broader community... The RFC is really vague. The unstable book only has a link to this issue. This issue mostly has detailed technical discussions. And I can't find that much info anywhere about what's changed or how stuff works.

I don't mean to complain, and I really truly appreciate all the hard work by those implementing, but I would also appreciate more transparency on this.

**Comment by alexreg** [comment]
@marcbowes Yeah, I'm kind of worried too. It seems like there's quite a lot of work left for stabilisation this year. I offered to work on opt-out hygiene for identifiers myself, but have received no response yet... Some greater transparency would be nice, as you say.

**Comment by petrochenkov** [comment]
Macros 2.0 are not even in the RFC stage yet - the @jseyfried's [RFC](https://github.com/jseyfried/rfcs/blob/hygiene/text/0000-hygiene.md) linked in the issue was never submitted, there's only very high level [RFC 1584](https://github.com/rust-lang/rfcs/pull/1584).
The implementation is purely experimental and mostly unspecified, and large part of it (not related to hygiene) is reused from `macro_rules!` without fixing problems that macros 2.0 were supposed to fix.

>Some greater transparency would be nice

The problem is that no work happen right now, so there's nothing to reveal :(
@nrc is busy, @jseyfried is busy, I worked on macros 2.0 a bit and would really like to dig into them more, but I'm busy too, unfortunately.
We need someone to adopt the feature, support and extend it, and gain expert-level knowledge of it, otherwise we will end up breaking hygiene and syntax details left and right after stabilization.

**Comment by alexreg** [comment]
@petrochenkov Ah, fair enough. I mean, that's a shame, but it makes sense at least. I guess this is an open call to anyone who might be able to take ownership of this feature, since no one comes to mind? I could still have a go at a little sub-feature, but I'm certainly in no position to take ownership of this.

**Comment by petrochenkov** [comment]
@alexreg 
>I could still have a go at a little sub-feature

Yes, please do!
Hygiene opt-out is one of the primary missing parts and implementing it would be useful in any case.

IIRC, two things need to be decide during implementation:
- Syntax for "unhygienic" identifiers (@jseyfried tentatively suggested `#ident` in https://github.com/rust-lang/rust/pull/40847)
- What exactly hygienic context the identifier introduced with `#ident` in a macro `m` will have - context of `m`'s invocation? context after expanding all macros ("no hygiene")? something else?

**Comment by alexreg** [comment]
> Syntax for "unhygienic" identifiers (@jseyfried tentatively suggested #ident in #40847)

Yeah, this was the plan. :-)

> What exactly hygienic context the identifier introduced with #ident in a macro m will have - context of m's invocation? context after expanding all macros ("no hygiene")? something else?

What's your inclination? I'm leaning towards the context of m's invocation, but curious to hear your thoughts...

**Comment by petrochenkov** [comment]
@alexreg 
>I'm leaning towards the context of `m`'s invocation

I think this is what should be implemented first, just because this is a more conservative alternative.

On the other hand, it makes writing internal helper macros harder, e.g.
```rust
macro m_helper() {
    struct #S;
}

macro m() {
    m_helper!(); // `S` has this context
    let s = S; // OK
}

fn main() {
    m!(); // `S` is not accessible here
    let s = S; // ERROR
}
```

**Comment by alexreg** [comment]
@petrochenkov Yeah, good point. I wonder if adding syntax like `m_helper!#()` (reuse the current context for the invocation) would help with that, or if there's a more elegant way that covers all use cases without too much pain?

**Comment by pierzchalski** [comment]
At that point you might start providing utility macros like `lift!(m_helper!(...))` (move all tokens generated by `m_helper!(...)` up one context, essentially pretending `foo!` was called by `main` in the example above):

```rust
macro m_helper() {
    struct #S;
    struct T;
}

macro m() {
    lift!(m_helper!()); // Sets caller context of `m_helper!` to caller context of `m!`.
    let s = S; // Not OK: `S` is in callers context.
    let t = T; // Not OK: `T` is in `m_helper!` context.
}

fn main() {
    m!();
    let s = S; // OK: `S` is in `main` context.
}
```

As a bonus, I think this would be possible to implement with some minor extensions to the proc macro API (mostly getting and setting the parent of an arbitrary scope/span, rather than only having access to the def and call site scopes).

**Comment by alexreg** [comment]
@pierzchalski Yeah, that's not a bad idea at all. Thoughts, @jseyfried / @petrochenkov?

**Comment by pierzchalski** [comment]
@alexreg Actually, I just realised this is the use-case I was looking for for `call_from` in the proc macro [RFC](https://github.com/rust-lang/rfcs/pull/2320) I put up.

**Comment by alexreg** [comment]
@pierzchalski I'll give that RFC a read tomorrow. Anyway, I certainly won't be including this `lift` macro or similar into my RFC; only the basic hygiene opt-out syntax. If the `lift` macro could go in another crate eventually, that would be ideal.

Incidentally, you seem to have a good knowledge of macro expansion and hygiene. Could I tempt you to contribute to <https://github.com/rust-lang-nursery/rustc-guide/issues/15>? :-)

**Comment by pierzchalski** [comment]
@alexreg Unfortunately my understanding of expansion and hygiene is a bit abstract - you'll notice the reference-level explanation in the RFC was rather thin! But if the RFC is accepted and I end up implementing it then I'd definitely like to document what I discover along the way.

**Comment by dtolnay** [comment]
I filed #49629 to consider dropping support for `#[$m:meta]` in favor of `#[$($meta:tt)*]` now that attributes are allowed to contain an arbitrary token stream.

**Comment by isiahmeadows** [comment]
Stylistic question: why is it `macro foo()` and not `macro foo!()`, like the way it's called?

**Comment by mark-i-m** [comment]
I believe the question mark isn't considered part of the name.

**Comment by isiahmeadows** [comment]
@mark-i-m You mean exclamation point, not question mark, right?

Also, in the docs, it usually refers to macros including the exclamation point as if it *were* part of the name, such as `println!` or `vec!`. As a concrete example, [here's one page in the book that uses it exclusively](https://doc.rust-lang.org/book/2018-edition/appendix-04-macros.html).

**Comment by mark-i-m** [comment]
Oh, yes, I meant exclamation mark. I might be wrong, but I believe the compiler itself doesn't count the exclamation mark as party of the ident.

**Comment by isiahmeadows** [comment]
@mark-i-m I would expect the *compiler* not to, but I'm speaking of the *language itself*, not the *implementation*. I'm suggesting matching what people think, not what computers process.

**Comment by mark-i-m** [comment]
That said, I don't see any reason why we cannot do something different from the internal representation.

If we did want to make ! part of the name, we would also want to do it in macro imports in the 2018 edition.

**Comment by isiahmeadows** [comment]
@mark-i-m If you change how you do it in imports, you could even integrate them into `use`, to avoid the need to deal with separate syntax altogether:

```rs
use mod::foo::some_macro!;
use mod::foo::{some_fn, some_macro!};
```

Even as recently as a couple days ago, I had to google how to import a macro from a peer module, and this would make it a million times simpler. It also wouldn't require any special attribute or whatever - it'd just work.

Similarly, exporting macros could be simply `pub macro foo!()` or similar. Alternatively, because `!` could serve as a delimiter, you could remove `macro` altogether and just do `pub fn foo!($a: ident) { ... }`, or you could reserve `macro foo!` for macros with multiple syntax rules (like [`vec!`](https://doc.rust-lang.org/std/macro.vec.html)), and let `fn foo!` be for ones with only a single value (like how [`try!`](https://doc.rust-lang.org/src/core/macros.rs.html#299-307) could be written if [this RFC](https://github.com/rust-lang/rfcs/blob/master/text/2298-macro-at-most-once-rep.md) gets implemented.

**Comment by mark-i-m** [comment]
@isiahmeadows you can do `use foo::bar::some_macro` in the 2018 edition. It is part of the module/macro system changes being stabilized.

Long term IIUC, the plan is to get rid of the attributes altogether (currently, you still need them to export macros).

I strongly prefer not to conflate functions and macros. They are very different.

**Comment by dhardy** [comment]
> `use mod::foo::{some_fn, some_macro!};`

The `!` suffix is not technically necessary here, but since this may be the only thing pointing out that it is indeed a macro, seems like a good idea (if feasible and not too late — probably is too late).

But of course if `foo` and `foo!` are two distinct identifiers, they should not both be allowed to exist in the same scope.

**Comment by mark-i-m** [comment]
Macros live in another namespace. You can have a macro `foo` and a function `foo` and a type `foo` all coexisting.

**Comment by Ekleog** [comment]
This makes me wonder. If I have a crate `mycrate` defining function `foo` and procedural macro `foo`, then if I do `use mycrate::foo`, which one is imported into scope? both? (I'm starting from the idea that some day in the hopefully near future the `-derive` crates and the not-`-derive` crates will be merge-able)

Adding the `!` as something to remove the ambiguity would likely help (in addition to being more explicit about what happens, which is always a good thing).

Assuming it's still time to do it, though… otherwise, add the `!` suffix as a possibility and a warning to push using it?

**Comment by Nemo157** [comment]
@Ekleog yes, both (along with potentially a module `foo` and type `foo`  as well).

**Comment by SimonSapin** [comment]
> If I have a crate mycrate defining function foo and procedural macro foo

You cannot do that today:

```
error: `proc-macro` crate types cannot export any items other than functions tagged with `#[proc_macro_derive]` currently
 --> src/lib.rs:1:1
  |
1 | pub fn some_public_function() {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

However in an hypothetical future Rust where that restriction is lifted somehow, or if a non-proc-macro crate re-exports a proc macro and also a function of the same name, then yes `use` will import both.

**Comment by Nokel81** [comment]
Should small proposals for Macros2.0 be brought up here or in a separate issue?

**Comment by GrayJack** [comment]
What the status of this?

**Comment by petrochenkov** [comment]
>What the status of this?

Bugs in macros 2.0 are occasionally fixed, but the feature is not on the 2019 roadmap.

**Comment by johnw42** [comment]
I'd like to clarify the rules about visibility and propose a mechanism for breaking hygiene in a controlled way.

I believe visibility should be resolved relative to where the macro is defined. For example, imagine a module like this:

    fn foo<T>(arg: T) {...}
    pub macro call_foo($arg:expr) => {
        foo($arg)
    }

With the current rules, expanding `call_foo!` in a different module doesn't work because `foo` isn't public. It should be should be fine because `foo` is visible where `call_foo!` is defined.

As for breaking hygiene, I think some additional flexibility would be very useful, and I think some solution needs to be introduced at the same time as the new hygiene rules, because otherwise it becomes impossible to break hygiene in ways that macros can depend on now (e.g. when defining a new type). I propose a built-in macro whose argument is an identifier that will appear verbatim in the expansion of a macro where it's used. Let's call it `verbatim!` for now (although I don't think that's a great name for it). It could, for example, be used to define an "anaphoric" map whose argument is an implied closure with a fixed argument name:

    pub macro map_it($iter:expr, $body:expr) {
        $iter.map(|verbatim!(it)| $body)
    }

It could also be used to translate an old macro that defines a type in the current module:

    // before:
    macro_rules! define_foo {
        () => {
            struct Foo {...}
        }
    }
    
    // after:
    pub macro define_foo() => {
        struct verbatim!(Foo) {...}
    }

Special names like `self` and `Self`, which are treated as identifiers for the purpose of macro expansion, should be implicitly verbatim. With the current rules, this doesn't work because `$body` can't refer to `self`:

    macro_rules! paranoid_method {
        ($name:ident, $body:expr) => {
            pub fn $name(&mut self) {
                self.verify_preconditions();
                $body;
                self.verify_postconditions();
            }
        }
    }

Currently the only workaround is to pass `self` as an additional argument to the macro, which is pretty silly since passing any other identifier would produce a syntactically invalid expansion.

One last thing on my wishlist is to extend `verbatim!` to support multiple arguments which are concatenated together to produce a new verbatim identifier. This could be used to do something like define a pair of related methods:

    /// Defines a pair of conversion methods, `as_$name` and `into_$name`,
    /// where `into_$name` consumes `self` and `as_$name` clones `self`.
    macro_rules! conversions {
        ($name:ident, $ty:ty, $body:expr) => {
            pub fn verbatim!(into_, $name)(self) -> $ty { $body }
            pub fn verbatim!(as_, $name)(&self) -> $ty {
                self.clone().verbatim!(into_, $name)
            }
        }
    }

**Comment by spearman** [comment]
> I'd like to clarify the rules about visibility and propose a mechanism for breaking hygiene in a controlled way.
> 
> I believe visibility should be resolved relative to where the macro is defined. For example, imagine a module like this:
> 
> ```
> fn foo<T>(arg: T) {...}
> pub macro call_foo($arg:expr) => {
>     foo($arg)
> }
> ```
> 
> With the current rules, expanding `call_foo!` in a different module doesn't work because `foo` isn't public. It should be should be fine because `foo` is visible where `call_foo!` is defined.

Could you give an example of this? This seems to work: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=a3e26640f0fb170b9a5a620250155f1c

You might need an absolute path `$crate::path::to::foo($arg)`, but I'm not sure if this applies to decl macros or only macro_rules.

**Comment by jjpe** [comment]
> It could, for example, be used to define an "anaphoric" map

I know it's just meant as an example, but anaphoric macros were a bad idea in Common Lisp, and I believe they're a bad idea here. The issue with them is that they don't scale: imagine that you end up in a situation where you have a nested usage of `map_it`. What does `it` refer to? 
Of course here it's easy enough to figure out in this contrived example, but at scale it causes ambiguity in the mind of the programmer.

> `verbatim!(it)`
I would like to note that that requires macros to be expandable to identifiers. I don't believe they currently have that capacity, so that would need to be added as well.
As for the name, I believe this is one instance where another cue from Lisps might help. While they use `~` (Closure) or `,` (CL) for interpolation, we could similarly use a sigil for identifier interpolation (leaving the default case to be verbatim instead).

**Comment by mayabyte** [comment]
Weather or not anaphoric macros are allowed, the ability to create new identifiers from ones passed into the macro seems highly desirable. Consider the following case with `macro_rules!`:
```rust
macro_rules! example {
    ($($a:ident),*) => {
        $(
            // assume each $a is an identifier for a HashMap, where each $a can have a different type
            let side_effect = $a.remove("some_key");
            ... // do some other stuff with side_effect
        ),*
        ...
        $( 
            // do some more things with each side effect created earlier, like putting them in a tuple
        ),*
    };
}
```

This doesn't work because `side_effect` has to have a fixed name, so it'd be shadowed each iteration. (Achieving something like this is possible, but it's hacky and involves a lot of boilerplate).

Being able to do something like one of these:
```rust
let side_effect_${a} = ...
let verbatim!(side_effect_, $a) = ...
```
would be really nice.

**Comment by eddyb** [comment]
Doesn't `side_effect_$` already successfully tokenize as `side_effect_ $`?

**Comment by petrochenkov** [comment]
Some status update (copypasted from Zulip https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/decl.20macro.20syntax/near/194354354):

`macro` items have multiple components to figure out before stabilizing, syntactic and semantic:

- `Span::def_site`, pretty far away, requires implementing cross-crate hygiene *at least*, and then formalizing stuff in `hygiene.rs` more carefully, in application to type-relative paths in particular.
- Syntax of the macro's left side (macro "parameters"), requires major design work to figure out future-compatibility (currently done with FIRST and FOLLOW sets) and figuring out how to match or not match all arms at the same time.
- Syntax of the macro's right side (macro body), requires a syntax for hygiene opt-out perhaps, maybe simplifying the use of repetitions (https://github.com/rust-lang/rust/issues/61053#issuecomment-506950747), but otherwise seems ok, it's just an arbitrary token stream.
- Surface syntax of the macro. The last year I almost wrote and RFC to set the top-level syntax `macro single_arm() {}` + `macro multiple_arms { (lhs1) {rhs1} (lhs2) {rhs2} }` in stone, but them recalled that people wanted the macros want to control what delimiters they are invoked with (e.g. restrict `vec![]` to only use square brackets), and that added more questions to the surface syntax, and I didn't write anything.

**Comment by petrochenkov** [comment]
One more issue is expanding fragments like `$e:expr` as token streams rather than AST pieces (this is also mentioned in the top comment and links to https://github.com/rust-lang/rust/issues/26361), but this is equally applicable to `macro_rules` where it should be doable in a (almost) backward-compatible way.

**Comment by jgarvin** [comment]
Do macros 2.0 address the self both is and isn't an identifier inconsistencies described here?: https://danielkeep.github.io/tlborm/book/mbe-min-non-identifier-identifiers.html 

Or the confusing behavior of macro_rules invoking other macro_rules macros not behaving the same as calling directly?:
https://danielkeep.github.io/tlborm/book/mbe-min-captures-and-expansion-redux.html

**Comment by petrochenkov** [comment]
>Do macros 2.0 address the self both is and isn't an identifier inconsistencies described here?

`self` is always an identifier, keywords are a (reserved) subset of identifiers.
I don't think macros 2.0 change anything here.

>Or the confusing behavior of macro_rules invoking other macro_rules macros not behaving the same as calling directly?

This is not decided and needs design.

**Comment by jgarvin** [comment]
@petrochenkov are you saying the the author is mistaken, or that rust has changed since he wrote it?

**Comment by petrochenkov** [comment]
@jgarvin 
Neither. The author is right in the sense that all the examples are correct etc., but he uses a different definition of "identifier" than the language and that's probably the source of the confusion.
(Anyway, this is pretty off-topic for this issue.)

**Comment by A1-Triard** [comment]
Does macros 2.0 allow to parse generic parameters definition?
With current `macro_rules!` the best approximation is `< $( $lt:tt $( : $clt:tt $(+ $dlt:tt )* )? ),+ $(,)?>`, but it does not cover all possible cases.

**Comment by mexus** [comment]
Does anybody happen to know if there are any plans to put "macros 2.0" onto the 2021 roadmap?

**Comment by mark-i-m** [comment]
Given the ruat 2021 posts I've read so far and my vague knowledge of the state of things, it seems unlikely. There is still some design work needed and there doesn't seem to be anyone interested in pushing it over the finish line atm.

**Comment by IQBigBang** [comment]
This is maybe out of scope of this document, but now that Rust has (or will soon have) constant functions and miri enabling it to run rust code at compile-time, wouldn't it be possible to support in-crate procedural macros which don't work essentialy as a compiler plugin? This seems like something I personally would expect a new macro system to enable.

**Comment by spearman** [comment]
Private decl macro shows up in generated docs (`rustc 1.56.0-nightly (0035d9dce 2021-08-16)`)
```
// src/lib.rs
#![feature(decl_macro)]
#[derive(Clone, PartialEq)]
pub struct Foo;
macro macro1 ($foo:ty) {
  impl Eq for $foo { }
}
macro_rules! macro2 {
  ($foo:ty) => {
    impl Copy for $foo { }
  }
}
macro1!(Foo);
macro2!(Foo);
```
`macro1` is shown in generated `cargo doc`, `macro2` is not.

**Comment by eddyb** [comment]
Before being reminded that `Span::def_site()` on the proc macro side is what hygiene is tracked under, I came up with this silly little snippet:
```rust
macro_rules! foo_impl {
    // ...
}

macro_rules! define_foo_with_hygiene {
    ($hygienic_crate:ident, $hygienic_foo_impl:ident) => {
        #[macro_export]
        macro_rules! foo {
            ($($inputs:tt)*) => {
                $hygienic_foo_impl!($hygienic_crate; $($inputs)*);
            };
        }
    };
}

macro define_foo() {
    define_foo_with_hygiene!($crate, foo_impl);
}
define_foo!();
```

The idea there is that by using `$crate` from inside a `macro` (as opposed to `macro_rules!`), the "hygiene 2.0" aspect was captured, and private parts of `$crate` can now be accessed through it (in a kind of objcap-y way, heh).
(Similarly, passing `foo_impl` from the `macro` would allow `foo_impl!` to remain unexported)

And you would only need to stabilize `"macro" IDENT "(" ")" "{" TOKEN_TREE* "}"` syntax, completely obscuring any further changes to MBE input LHS pattern syntax, parsing/choice, or RHS expansion.

But if a proc macro can use `Span::def_site()` to achieve the same thing, this trick is probably less useful.

**Comment by danii** [comment]
Should metavariables be prevented from being named `self` or as keywords? I feel like metavariable keywords could be used for future expansions to macros, like associated macros with `$self`.

**Comment by danielrab** [comment]
yeah, I agree, no keywords should probably be valid names for metavariables, in case we want to make any of them special somehow

**Comment by eap314** [comment]
These suggestions might be out of scope for this discussion, but it beats the RFC process, so
- hygiene could be (mostly) achieved by having macros prefix functions with h_ and normal functions be prefixed uh_
- If you don't want hygiene, you could use a keyword like `unhygienic` to prefix with uh_
```
unhygienic macro count {}
macro count2 {}

count!();
unhygienic { count2!(); }
```

To bring up the out of scope part,

- Proc macros can be preprocessed, removing the need for a separate crate
- You can `proc use` a crate to bring it into scope during preprocessing

**Comment by A1-Triard** [comment]
>     * hygiene could be (mostly) achieved by having macros (mangle) prefix functions with h_ and normal functions be (mangle) 

As far as I understand it would not work for `#[no_mangle]` functions. If this is correct, then it means the suggestion just does not work.

**Comment by lygstate** [comment]
When this will be stablized :)

**Comment by clarfonthey** [comment]
Presumably it will be stabilised *after* all of the requisite tasks in the issue description are completed.

**Comment by cybersoulK** [comment]
i may not be understanding macros 2.0 right, but these are the benefits i expect:

1 - I can use crates, Types and functions that i define in my library directly without $crate. (deprecate $crate)
2 - All of the items created in the macro should be expanded (that's the whole point of macros).

This is logical:
- if you want Types/functions to be private, just define them outside the macro, where the macro can access them (1)
- Macro expansion can be read like normal functions in the library, and the tokens are the only inputs from user library.
- It's up to the user where the macro expands. He can wrap it with a mod {}, or limit the scope with a block {}.

**Comment by cybersoulK** [comment]
I think the confusion we are facing now is how to deal when the Library code mixes the User code, and it's expected to all mix with macros, since by definition, it's the library code blueprint being copy pasted to the user scope.
Perhaps better tooling is required on how to handle when two Textual words collide, instead of trying to avoid the problem and make all kinds of limitations to macros.

**Comment by jjpe** [comment]
@cybersoulK do everyone here a favor please, and don't multipost. Everyone who is subscribed to this gets a *separate* email about each post, and the signal:noise ratio is dropping real fast ATM.

**Comment by dengyunsheng250** [comment]
Has there been any progress on Macros 2.0 lately?

**Comment by jhpratt** [comment]
No, but there will hopefully be a general increase in knowledge in what the issues are, as the proposal for a macros working group has been accepted. The WG will be created shortly!

**Comment by tgross35** [comment]
It could be interesting to allow a way to reverse parsing to match from right to left.  This can be faked using an accumulator and recursively popping tokens until you get a match on a single one, but this is a pretty messy pattern. A more simple way to create a reverse TT muncher would make it much easier to apply precedence to contents, such as when evaluating math.

```rs
// Works
macro first {
    ($first:tt $($rest:tt)*) => { println!("first item is {}", $first) };
}

// Fails to compile (multiple parsing options)
macro last {
    #[reverse] // maybe this could make it work
    ($($rest:tt)* $last:tt) => { println!("last item is {}", $last) }
}
```

**Comment by clarfonthey** [comment]
The main issue with RTL parsing is that, at least for now, parsing doesn't actually store the tokens being returned in a way that lets you iterate backwards. You could allow this and do it without allocating if you permit parsing the tokens twice (once to find the end of the macro invocation, the second time to do the actual parsing), but it's still a performance penalty to allow parsing from the other direction.

Of course, the usual way, recursively expanding an intermediate macro in reverse order, is probably slower than offering a dedicated method that allows this. If this were done, it would be nice to also have `proc_macro::TokenStream` implement `DoubleEndedIterator` to reflect that, and allow proc macros to also take advantage of that.

**Comment by Ygg01** [comment]
Here is to hoping two big issues are getting solved.

I tried solving macro hygiene but one pre-requisite was learning how other languages do it. Learning Racket is where I lost my motivation.

**Comment by petrochenkov** [comment]
> Here is to hoping two big issues are getting solved (hygiene bending and nesting macros).
> 
> I tried solving macro hygiene but one pre-requisite was learning how other languages do it. Learning Racket is where I lost my motivation.

The issue about hygiene is ["Tracking issue for `Span::def_site()`"](https://github.com/rust-lang/rust/issues/54724), this issue is blocked on it.
In general, declarative macros are a syntactic sugar for procedural macros, so the "procedural macros 2.0" need to be implemented first.

Racket needs to be learned just enough to understand https://users.cs.utah.edu/plt/scope-sets/ in detail, to be able to possibly tweak it to match Rust realities.

**Comment by lolbinarycat** [comment]
one usecase of macros where rust does worse than C is local unhyginic macros used in the definition of a single array or function.

some way to bind specific variables to to those from the outer scope, or to disable macro hygine for certain identifies, would be appreciated.

perhaps i need to call a function 10 times, and each time i need to pass 7 different arguments, 6 of which are always the same local variables, and only the 7th changes.  C could easily manage this with a define and undef, but in rust, you would have to pass all the local variables to each macro call, which wouldn't actually simplify the code at all.

**Comment by felix91gr** [comment]
> which wouldn't actually simplify the code at all.

That is true for this version of "simplify": 

Simpler code: *shorter code.*

But it is not true for the following version of "simplify":

Simpler code: *code that is easier to understand and maintain*.

I don't think locally disabling hygiene helps at all, if what we seek is simpler code in the second sense.

**Comment by ifsheldon** [comment]
> one usecase of macros where rust does worse than C is local unhyginic macros used in the definition of a single array or function.

I think "worse" here is generally better. I've read a pretty bad C codebase once, which is full of unhygine macros that means a lot of sloppy auto copy-and-paste. IDEs could not even help me to analyze the code. I really appreciate Rust didn't just choose the simple way to do macros.

> perhaps i need to call a function 10 times, and each time i need to pass 7 different arguments, 6 of which are always the same local variables, and only the 7th changes.

Make a closure please.

**Comment by tmccombs** [comment]
> Make a closure please.

A closure can't do everything a macro can.  For example it can't return from the parent function.

**Comment by mayabyte** [comment]
> perhaps i need to call a function 10 times, and each time i need to pass 7 different arguments, 6 of which are always the same local variables, and only the 7th changes.

I think what you're describing is [partial application](https://en.m.wikipedia.org/wiki/Partial_application) which is something that can already be achieved with existing crates. Combining this with the Try operator could achieve the early-return functionality described above as well.

**Comment by lolbinarycat** [comment]
it turns out the existing `macro_rules!` macros can already access identifiers from their parent scope, so this type of local macro actually already works, for when you need to do something that partial application can't do.

**Comment by oli-obk** [comment]
Tracking issues are not intended as places for discussions. This issue already has over 100 comments and lots of subscribers. Please open separate issues and link them, instead of having the discussion here.

**Comment by safinaskar** [comment]
🚀 I found a way to evaluate `concat_idents` (and `concat` and few other built-in macros) **before** evaluating other macro, which takes `concat_idents` as an argument! I. e. I found a way to evaluate `a!(concat_idents!(...))` such way, that `concat_idents` evaluates before `a`. Answer is crate https://crates.io/crates/with_builtin_macros !!! Thanks, @danielhenrymantilla ! In other words, `with_builtin_macros` is `paste`, but not only for `concat_idents`, but also for `concat` and some other macros.

And in other words, `with_builtin_macros` allows one to achieve eager evaluation of macros in limited way.

Also, `with_builtin_macros` allows one to use `concat_idents` when defining new identifier.

Also, https://crates.io/crates/with_builtin_macros allows one to use `concat_idents` in stable Rust.

```rust
// (This code was not tested, may contain typos)

fn concat_idents!(a, b) () {} // Doesn't work

with_builtin_macros::with_eager_expansions! {
  fn #{ concat_idents!(a, b) } () {} // Works! Even on stable!
}

macro_rules! this_macro_accepts_ident {
  ($a:ident) => {}
}

// Doesn't work, because "this_macro_accepts_ident" evaluates before "concat_idents"
this_macro_accepts_ident!(concat_idents!(a, b));

with_builtin_macros::with_eager_expansions! {
  this_macro_accepts_ident!(#{ concat_idents!(a, b) }); // Works! Even on stable!
}

macro_rules! this_macro_accepts_literal {
  ($a:literal) => {}
}

// Doesn't work.
// Moreover, you cannot solve this problem using #[feature(macro_metavar_expr_concat)],
// because ${concat(...)} produces identifier, not string literal!!!
// Same applies to "paste"! "paste::paste!" deals with identifiers, not strings. So, with_builtin_macros is the only way!!!
this_macro_accepts_literal!(concat!("a", "b"));

with_builtin_macros::with_eager_expansions! {
  this_macro_accepts_literal!(#{ concat!("a", "b") }); // Works! Even on stable!
}
```

**Comment by tgross35** [comment]
@safinaskar you have posted the same comment on three different tracking issues related to macros - providing stable alternatives can be helpful, but once on the most relevant issue is enough. The suggestion also isn't at all related to this tracking issue's feature.

**Comment by safinaskar** [comment]
> you have posted the same comment on three different tracking issues related to macros

Yes. Because I spent a lot of time searching for solution. And so I posted it everywhere, so everybody will be able to easily find it

> providing stable alternatives can be helpful

In some cases `with_builtin_macros` is not just "alternative", but the only solution. E. g. when you need to eagerly evaluate `something!(concat!(a, b))`

But okay, I will post to one issue only next time

**Comment by max-kamps** [comment]
Not sure if this is known (or even an intentional limitation?), but for some reason macros can't refer to type parameters or const parameters at their definition site:

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=fee4172d5f73a52e62614e77a3546d72

```rs
#![feature(decl_macro)]
use std::fmt::Debug;


fn test<TypeParam: Default + Debug, const CONST_PARAM: usize>() {
    #[derive(Default, Debug)]
    struct LocalType(usize);
    
    macro can_refer_to_local_type() { dbg!(LocalType::default()) }
    can_refer_to_local_type!();
    
    macro cant_refer_to_type_param() { dbg!(TypeParam::default()) }
    cant_refer_to_type_param!();
    // Error: failed to resolve: use of undeclared type `TypeParam`
    
    
    const LOCAL_CONST: usize = 812;

    macro can_refer_to_local_const() { dbg!(LOCAL_CONST) }
    can_refer_to_local_const!();
    
    macro cant_refer_to_const_param() { dbg!(CONST_PARAM) }
    cant_refer_to_const_param!();
    // Error: cannot find value `CONST_PARAM` in this scope
    // you might have meant to refer to this const parameter
}
```
